### PR TITLE
[python] Import somacore

### PR DESCRIPTION
**Pull request by johnkerl**
Purpose: Validates that the Python TileDB-SOMA implementation is following the SOMA spec at the class/function/signature level.

Context: #669

### CODE REVIEW

**Comment by johnkerl** [comment]
@thetorpedodog I could use some advice from you on CI errors that I don't understand:

```
src/tiledbsoma/sparse_nd_array.py:26: error: Class cannot subclass "SparseNDArray" (has type "Any")  [misc]
src/tiledbsoma/sparse_nd_array.py:308: error: Class cannot subclass "SparseRead" (has type "Any")  [misc]
src/tiledbsoma/dense_nd_array.py:23: error: Class cannot subclass "DenseNDArray" (has type "Any")  [misc]
src/tiledbsoma/dataframe.py:25: error: Class cannot subclass "DataFrame" (has type "Any")  [misc]
```

**Comment by thetorpedodog** [comment]
> ```
> src/tiledbsoma/sparse_nd_array.py:26: error: Class cannot subclass "SparseNDArray" (has type "Any")  [misc]
> src/tiledbsoma/sparse_nd_array.py:308: error: Class cannot subclass "SparseRead" (has type "Any")  [misc]
> src/tiledbsoma/dense_nd_array.py:23: error: Class cannot subclass "DenseNDArray" (has type "Any")  [misc]
> src/tiledbsoma/dataframe.py:25: error: Class cannot subclass "DataFrame" (has type "Any")  [misc]
> ```

I believe this happens because `somacore` is not installed in the type-checking environment, so mypy doesn’t know anything about the types there. If you install it there, that should do it, I think (may also solve your anndata problems?).

**Comment by johnkerl** [comment]
Explanation from @thetorpedodog on the CI fails:

```
src/tiledbsoma/sparse_nd_array.py:26: error: Class cannot subclass "SparseNDArray" (has type "Any")  [misc]
src/tiledbsoma/sparse_nd_array.py:308: error: Class cannot subclass "SparseRead" (has type "Any")  [misc]
src/tiledbsoma/dense_nd_array.py:23: error: Class cannot subclass "DenseNDArray" (has type "Any")  [misc]
src/tiledbsoma/dataframe.py:25: error: Class cannot subclass "DataFrame" (has type "Any")  [misc]
```

* The linter step does not have access to `somacore` (or `numpy` or `anndata` or anything else)
* For other packages like `numpy` this is OK as it infers types from packages it doesn't know about as `Any`, which often works
* For `somacore` it doesn't work as we're doing things like `class CollectionBase(TileDBObject, somacore.Collection[CollectionElementType]):` where the second argument is inferred as `Any`, and `Any` _cannot_ be subclassed
* As an experiment I tried https://github.com/single-cell-data/TileDB-SOMA/pull/678/commits/0b42d3bc5b8a03ca5c067ce6320dffdcaba1e07a but that (a) unleashed a torrent of new errors, and (b) didn't solve the existing ones

I do not know how to move forward with this PR.

cc @bkmartinjr @thetorpedodog @nguyenv @gsakkis -- I'm hoping someone with more Python expertise can see a way forward.

**Comment by johnkerl** [comment]
Update -- @thetorpedodog noticed
> tiledbsoma/dense_nd_array.py:5: error: Skipping analyzing "somacore": module is installed, but missing library stubs or py.typed marker
in the logs which I had missed & which is googleable -- I'll pick up from here tomorrow

**Comment by thetorpedodog** [comment]
Part of the problem should be resolved by `pip install`ing `somacore` in the mypy environment. Many (all?) of the other problems might be taken care of with some new type hints here: https://github.com/single-cell-data/TileDB-SOMA/pull/683

**Comment by gsakkis** [comment]
The path of least resistance for now is relaxing mypy strictness by adding `allow_subclassing_any = True` to the `[mypy]` section of `setup.cfg`. 

Making the necessary changes for strict type checking in the presence of dependencies (mainly Numpy) is a larger endeavor to be addressed in a separate PR.

**Comment by johnkerl** [comment]
@thetorpedodog @gsakkis with the most recent commit implementing @gsakkis 's excellent suggestion we have green CI! 🎆

**Comment by thetorpedodog** [comment]
If you add `somacore==0.0.0a2` to the packages installed when type-checking, it does reveal some genuine problems:

```
src/tiledbsoma/collection.py:59: error: Type argument "CollectionElementType" of "Collection" must be a subtype of "SOMAObject"  [type-var]
src/tiledbsoma/sparse_nd_array.py:50: error: Cannot override final attribute "soma_type" (previously declared in base class "SparseNDArray")  [misc]
src/tiledbsoma/sparse_nd_array.py:147: error: Cannot override final attribute "is_sparse" (previously declared in base class "SparseNDArray")  [misc]
src/tiledbsoma/sparse_nd_array.py:162: error: Signature of "read" incompatible with supertype "SparseNDArray"  [override]
src/tiledbsoma/sparse_nd_array.py:162: note:      Superclass:
src/tiledbsoma/sparse_nd_array.py:162: note:          def read(self, slices: Any, *, batch_size: BatchSize = ..., partitions: Optional[ReadPartitions] = ..., result_order: Any = ..., platform_config: Optional[Mapping[str, Any]] = ...) -> SparseRead
src/tiledbsoma/sparse_nd_array.py:162: note:      Subclass:
src/tiledbsoma/sparse_nd_array.py:162: note:          def read(self, coords: Optional[Sequence[Union[None, Union[int, slice], Sequence[int], ndarray[Any, dtype[integer[Any]]], Any]]] = ...) -> SparseNDArrayRead
src/tiledbsoma/sparse_nd_array.py:257: error: Signature of "write" incompatible with supertype "SparseNDArray"  [override]
src/tiledbsoma/sparse_nd_array.py:257: note:      Superclass:
src/tiledbsoma/sparse_nd_array.py:257: note:          def write(self, values: Union[Any, Any, Any, Any], *, platform_config: Optional[Mapping[str, Any]] = ...) -> None
src/tiledbsoma/sparse_nd_array.py:257: note:      Subclass:
src/tiledbsoma/sparse_nd_array.py:257: note:          def write(self, values: Union[Any, Any, Any, Any]) -> None
src/tiledbsoma/dense_nd_array.py:46: error: Cannot override final attribute "soma_type" (previously declared in base class "DenseNDArray")  [misc]
src/tiledbsoma/dense_nd_array.py:142: error: Cannot override final attribute "is_sparse" (previously declared in base class "DenseNDArray")  [misc]
src/tiledbsoma/dense_nd_array.py:144: error: Signature of "read" incompatible with supertype "DenseNDArray"  [override]
src/tiledbsoma/dense_nd_array.py:144: note:      Superclass:
src/tiledbsoma/dense_nd_array.py:144: note:          def read(self, coords: Any, *, batch_size: BatchSize = ..., partitions: Optional[ReadPartitions] = ..., result_order: Any = ..., platform_config: Optional[Mapping[str, Any]] = ...) -> Any
src/tiledbsoma/dense_nd_array.py:144: note:      Subclass:
src/tiledbsoma/dense_nd_array.py:144: note:          def read(self, coords: Tuple[Union[int, slice], ...], *, result_order: Literal['row-major', 'column-major', 'auto'] = ...) -> Any
src/tiledbsoma/dense_nd_array.py:229: error: Signature of "write" incompatible with supertype "DenseNDArray"  [override]
src/tiledbsoma/dense_nd_array.py:229: note:      Superclass:
src/tiledbsoma/dense_nd_array.py:229: note:          def write(self, coords: Any, values: Any, *, platform_config: Optional[Mapping[str, Any]] = ...) -> None
src/tiledbsoma/dense_nd_array.py:229: note:      Subclass:
src/tiledbsoma/dense_nd_array.py:229: note:          def write(self, coords: Tuple[Union[int, slice], ...], values: Any) -> None
src/tiledbsoma/dataframe.py:[49](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/3885431318/jobs/6629298550#step:7:50): error: Cannot override final attribute "soma_type" (previously declared in base class "DataFrame")  [misc]
src/tiledbsoma/dataframe.py:1[61](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/3885431318/jobs/6629298550#step:7:62): error: Return type "Sequence[str]" of "index_column_names" incompatible with return type "Tuple[str, ...]" in supertype "DataFrame"  [override]
src/tiledbsoma/dataframe.py:195: error: Signature of "read" incompatible with supertype "DataFrame"  [override]
src/tiledbsoma/dataframe.py:195: note:      Superclass:
src/tiledbsoma/dataframe.py:195: note:          def read(self, ids: Any, column_names: Optional[Sequence[str]] = ..., *, batch_size: BatchSize = ..., partitions: Optional[ReadPartitions] = ..., result_order: Optional[Any] = ..., value_filter: Optional[str] = ..., platform_config: Optional[Mapping[str, Any]] = ...) -> ReadIter[Any]
src/tiledbsoma/dataframe.py:195: note:      Subclass:
src/tiledbsoma/dataframe.py:195: note:          def read(self, *, ids: Optional[Sequence[Union[None, int, slice, Sequence[int], Any, Any, ndarray[Any, dtype[integer[Any]]]]]] = ..., value_filter: Optional[str] = ..., column_names: Optional[Sequence[str]] = ..., result_order: Optional[Literal['row-major', 'column-major', 'auto']] = ...) -> TableReadIter
src/tiledbsoma/dataframe.py:310: error: Signature of "write" incompatible with supertype "DataFrame"  [override]
src/tiledbsoma/dataframe.py:310: note:      Superclass:
src/tiledbsoma/dataframe.py:310: note:          def write(self, values: Union[Any, Any], platform_config: Optional[Mapping[str, Any]] = ...) -> None
src/tiledbsoma/dataframe.py:310: note:      Subclass:
src/tiledbsoma/dataframe.py:310: note:          def write(self, values: Any) -> None
```

along with noisy warnings about NDArrays and dtypes: https://github.com/single-cell-data/TileDB-SOMA/actions/runs/3885431318/jobs/6629298550

**Comment by johnkerl** [comment]
> along with noisy warnings about NDArrays and dtypes: https://github.com/single-cell-data/TileDB-SOMA/actions/runs/3885431318/jobs/6629298550

OK @thetorpedodog for sandbox I'll do that & try to fix what I find

For CI purposes moving forward though I don't know how to have our cake & eat it too ... :( ... how to get it to warn only on the things we want to warn on and not on the others ... maybe there is more `setup.cfg` magic suppress-config ... 🤔

**Comment by johnkerl** [comment]
@thetorpedodog this is passing format checks and unit test in my sandbox with https://github.com/single-cell-data/SOMA/pull/79 applied ... how should we proceed?

**Comment by thetorpedodog** [comment]
`Experiment` is not yet present in that version of `somacore`. However, because the way `Measurement` and `Experiment` in `somacore` work is very different from the way they are currently implemented here (`somacore` implements them as wrappers, not subclasses). I would avoid trying to use them at the moment and just maintain the current `tiledbsoma`-specific implementations, and submit this change in that state.

Once https://github.com/single-cell-data/SOMA/pull/78 is landed, I intend to mark that as a new version. You should then be able to replace `tiledbsoma.query.query.ExperimentAxisQuery` with `somacore.query.query.ExperimentAxisQuery` (while still otherwise maintaining the `tiledbsoma`-specific `Experiment` class).

This should improve as we restructure the creation/opening API (cf. https://github.com/single-cell-data/SOMA/issues/41) and the `somacore` and `tiledbsoma` implementations fully line up.

**Comment by thetorpedodog** [comment]
> Once [single-cell-data/SOMA#78](https://github.com/single-cell-data/SOMA/pull/78) is landed, I intend to mark that as a new version. You should then be able to replace `tiledbsoma.query.query.ExperimentAxisQuery` with `somacore.query.query.ExperimentAxisQuery` (while still otherwise maintaining the `tiledbsoma`-specific `Experiment` class).

Actually, I can handle that part (once this change is in); it makes sense for me to cut it over and ensure that I don’t introduce any new bugs there.

**Comment by johnkerl** [comment]
> I would avoid trying to use them at the moment and just maintain the current tiledbsoma-specific implementations, and submit this change in that state.

@thetorpedodog the changes I'm making on this PR are not optional and not just on a whim. I'm struggling to make forward progress on https://github.com/single-cell-data/TileDB-SOMA/pull/678 and I simply do not know how to do so without the changes I am proposing here.

**Review by thetorpedodog — COMMENTED** [review]
There are some things in somacore HEAD which are unavailable in v0.0.0a2. The changes relative to this review are in changeset 34b70788a23f92344cb2907be21bfcebd0e731f4 and can be cherry-picked onto here. With this, everything works great on somacore v0.0.0a2.

Once everything gets merged together, the overall change looks like this: https://github.com/single-cell-data/TileDB-SOMA/pull/716

**Comment by johnkerl** [comment]
ok @thetorpedodog all CI is currently green .....

File: apis/python/src/tiledbsoma/sparse_nd_array.py [hunk 1]
```diff
@@ -161,9 +162,8 @@ def nnz(self) -> int:
 
     def read(
         self,
-        coords: Optional[SparseNdCoordinates] = None,
-        # *,
-        # TODO: missing parameters
+        slices: Optional[SparseNdCoordinates] = None,
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/sparse_nd_array.py:165]: This is a note to myself and **not** a change that you need to make: we should probably make sure that we have `coords` vs. `slices` standardized within `somacore` (and then we can pick it up from there), and then fix the implementation up once we make that consistent.

File: .github/workflows/python-ci.yml [hunk 2]
```diff
@@ -16,9 +16,9 @@ jobs:
       fail-fast: false
       matrix:
         # TODO: restore Windows build once we have C++/libtiledbsoma integration supported there
-        os: [ubuntu-22.04, macos-12]
+        os: [ubuntu-22.04]
```
Comment (thetorpedodog) [inline .github/workflows/python-ci.yml:19]: You'll want to undo these changes I made for my own testing convenience.

File: apis/python/src/tiledbsoma/util_tiledb.py [hunk 3]
```diff
@@ -57,47 +57,59 @@ def to_tiledb_supported_array_type(x: T) -> T:
...
+        if x.hasnans:
+            raise ValueError(
+                "Categorical array contains NaN -- unable to convert to TileDB array."
+            )
+        return x.astype(bytes)
+
+    return x.astype("O")
+
+
+_MT = TypeVar("_MT", np.ndarray, pd.Series, sp.spmatrix)  # type: ignore[type-arg]
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/util_tiledb.py:107]: leftover from earlier work I was doing; I think this is the right thing to do:
```suggestion
_MT = TypeVar("_MT", np.ndarray, sp.spmatrix)  # type: ignore[type-arg]
```
Also the `Any -> Any` above might not be The Right Thing but it’s fine for now since it’s an internal utility function.

**Review by thetorpedodog — APPROVED** [review]
Two minor things from the changes I was doing (and one note for myself), though neither are functional changes that affect execution itself. Looks good!
